lib/index.js

```diff
@@ -136,7 +136,9 @@ PostHTML.prototype.process = function (tree, options) {
     }
 
     // (re)extend the object
-    api(result)
+    if (result.extendApi === undefined) {
```

Comment:
```js
if (!result.messages) {
   result = new API(result)
}
```

?

Reply:
Tomorrow the message will not be what then?

Reply:
extendApi is flag. messages container

Reply:
If `walk`, `match` or `messages` are `undefined` => the API was lost => needs to be re-extended. Which API property gets checked is ?neglectable => no *real* need for an extra property

Reply:
function Api(oldTree)
{
	this.messages = (oldTree && oldTree.messages) || []
}

var tree = ['tree']


F.call(tree)

tree.messages.push('1')

console.log(tree.messages)
var n
F.call(n = ['$$$$$$$$$'], tree)
console.log(n.messages)

Reply:
@voischev 👍 

```js
function Api(tree) {
  this.messages = (tree && tree.messages) || []
}

var tree = [ 'tree' ]

Api.call(tree)

tree.messages.push('1')

console.log('Tree')
console.log(tree)
console.log(tree.messages)

var n

Api.call(n = ['$$$$$$$$$'], tree)

n.messages.push('2')

console.log('\n\nnew Tree')
console.log(n)
console.log(n.messages)

// Logs
// ----------
// Tree
// [ 'tree', messages: [ '1' ] ]
// [ '1' ]
// 
// 
// new Tree
// [ '$$$$$$$$$', messages: [ '1', '2' ] ]
// [ '1', '2' ]
````

How does the implementation in `posthtml` look like ? :)